drivers/entropy/entropy_mspm0_trng.c

```diff
@@ -0,0 +1,300 @@
+#include <zephyr/irq.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/ring_buffer.h>
+
+/* TI Driverlib includes */
+#include <ti/driverlib/dl_trng.h>
+#include <ti/devices/msp/peripherals/hw_trng.h>
+
+#define TRNG_BUFFER_SIZE	CONFIG_ENTROPY_MSPM0_TRNG_POOL_SIZE
+#define TRNG_DECIMATION_RATE	CONFIG_ENTROPY_MSPM0_TRNG_DECIMATION_RATE
+#define TRNG_SAMPLE_SIZE	4
+#define TRNG_SAMPLE_GENERATE_TIME (1000000 * (32 * (TRNG_DECIMATION_RATE + 1)) \
+		/ (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / (config->clock_divider)))
+
+struct entropy_mspm0_trng_config {
+	TRNG_Regs *base;
+	uint8_t clock_divider;
+};
+
+struct entropy_mspm0_trng_data {
+	struct k_sem sem_lock;
+	struct k_sem sem_sync;
+	struct ring_buf entropy_pool;
+	uint8_t pool_buffer[TRNG_BUFFER_SIZE];
+	bool initialized;
+};
+
+static int wait_for_command_done(const struct entropy_mspm0_trng_config *config)
+{
+	uint32_t timeout = 3500;
+
+	while (!DL_TRNG_isCommandDone(config->base)) {
+		if (timeout-- <= 0) {
+			return -EIO;
+		}
+	}
+	return 0;
+}
+
+static void configure_trng_clock_divider(const struct device *dev)
+{
+	const struct entropy_mspm0_trng_config *cfg = dev->config;
+
+	DL_TRNG_setClockDivider(cfg->base, (DL_TRNG_CLOCK_DIVIDE)(cfg->clock_divider - 1));
+}
+
+static int run_startup_tests(const struct device *dev)
+{
+	const struct entropy_mspm0_trng_config *config = dev->config;
+	uint8_t dig_test_result, ana_test_result;
+	int ret;
+
+	/* Move TRNG to NORM_FUNC state */
+	DL_TRNG_sendCommand(config->base, DL_TRNG_CMD_NORM_FUNC);
+
+	ret = wait_for_command_done(config);
+	if (ret < 0) {
+		return -EIO;
+	}
+
+	DL_TRNG_clearInterruptStatus(config->base, DL_TRNG_INTERRUPT_CMD_DONE_EVENT);
+
+	/* Run digital self-test */
+	DL_TRNG_sendCommand(config->base, DL_TRNG_CMD_TEST_DIG);
+
+	wait_for_command_done(config);
+	if (ret < 0) {
+		return -EIO;
+	}
+
+	DL_TRNG_clearInterruptStatus(config->base, DL_TRNG_INTERRUPT_CMD_DONE_EVENT);
+
+	/* Check digital test results */
+	dig_test_result = DL_TRNG_getDigitalHealthTestResults(config->base);
+	if (dig_test_result != DL_TRNG_DIGITAL_HEALTH_TEST_SUCCESS) {
+		return -EIO;
+	}
+
+	/* Run analog self-test */
+	DL_TRNG_sendCommand(config->base, DL_TRNG_CMD_TEST_ANA);
+	wait_for_command_done(config);
+	if (ret < 0) {
+		return -EIO;
+	}
+
+	/* Check analog test results */
+	ana_test_result = DL_TRNG_getAnalogHealthTestResults(config->base);
+	if (ana_test_result != DL_TRNG_ANALOG_HEALTH_TEST_SUCCESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int configure_normal_operation(const struct device *dev)
+{
+	const struct entropy_mspm0_trng_config *config = dev->config;
+	uint32_t dummy_data;
+
+	/* Clear IRQ_CAPTURED_RDY_IRQ status */
+	DL_TRNG_clearInterruptStatus(config->base, DL_TRNG_INTERRUPT_CAPTURE_RDY_EVENT);
+
+	/* Set decimation rate and send NORM_FUNC command */
+	DL_TRNG_setDecimationRate(config->base, (DL_TRNG_DECIMATION_RATE)TRNG_DECIMATION_RATE);
+
+	DL_TRNG_sendCommand(config->base, DL_TRNG_CMD_NORM_FUNC);
```

Comment:
Not sure if the whole idea of the driver is correct. If I understand correctly, From TRM `NORM_FUNC` means that the TRNG components will be turned on and considering the below interrupt enablement, ISR will be fired for every `DATA_CAPTURE` whether there is a application or subsys consumer called synchronous get_entropy or random from ISR of other drivers.